src/names.jl

```diff
@@ -0,0 +1,23 @@
+using NamedDims
+
+# Re-wrap with names:
+kronecker(A::NamedDimsArray{L1}, B::NamedDimsArray{L2}) where {L1,L2} =
+    NamedDimsArray(KroneckerProduct(parent(A), parent(B)), kron_names(L1, L2))
+
+kronecker(A::NamedDimsArray{L}, B::AbstractMatrix) where {L} =
+    NamedDimsArray(KroneckerProduct(parent(A), B), kron_names(L, (:_, :_)))
+kronecker(A::AbstractMatrix, B::NamedDimsArray{L}) where {L} =
+    NamedDimsArray(KroneckerProduct(A, parent(B)), kron_names((:_, :_), L))
+
+# Power
+kronecker(A::NamedDimsArray{L}, p::Int) where {L} =
+    NamedDimsArray(kronecker(parent(A), p), kron_names(L, Val(p)))
+
+# Finding the names
+kron_names(left::Tuple, right::Tuple) = map(_join, left, right)
+
+_join(i::Symbol, j::Symbol) = _join(Val(i), Val(j))
+@generated _join(::Val{i}, ::Val{j}) where {i,j} = QuoteNode(Symbol(i, :⊗, j))
```

Comment:
I think using `:⊗` is not appropriate in this context. Maybe just use the more neutral `-` to join symbols?

Reply:
The problem is that anything which parses as infix gets displayed (and has to by typed) in an ugly way:
```
julia> Symbol(:a, :-, :b)
Symbol("a-b")

julia> Symbol(:a, :×, :b)
Symbol("a×b")

julia> Symbol(:a, ",", :b)
Symbol("a,b")

julia> Symbol(:a, :_, :b)
:a_b
```
How do you feel about an underscore?

Reply:
Like so:
```
julia> nda = NamedDimsArray(rand(Int8,2,2), (:x, :y));

julia> kronecker(nda, 2)
4×4 NamedDimsArray{(:x_x, :y_y),Number,2,KroneckerPower{Array{Int8,2},2}}:
 105    82    82   -92
 -41  -123    46  -118
 -41    46  -123  -118
 -23   -69   -69    49

julia> kronecker(nda, 4)
16×16 NamedDimsArray{(:x_x_x_x_x_x_x_x, :y_y_y_y_y_y_y_y),Number,2,KroneckerPower{Array{Int8,2},4}}:
```
Or should that be `(:x_4, :y_4)` perhaps?

(Also, why `Number` here?)

Reply:
Underscores seem fine.

The `Number` seems to be a mistake. The idea was that it contained the order of the Kronecker power (how many matrices are used in the product) so that this is available to the compiler. In the constructor, it seems to add the type of the order rather than the value. Should fix this or remove altogether...

Reply:
Done! 

Not sure I follow, `eltype(kronecker(rand(Int32,2,2), 2)) == Int32` seems fine, but `eltype(kronecker(nda, 2)) == Number` seems like a bug?

Reply:
The only weirdness now is that `:_` is also used as a wildcard for ordinary arrays, thus:
```
julia> kronecker(nda, rand(2,2), nda, rand(2,2))
16×16 NamedDimsArray{(:x___x__, :y___y__),Number,2,Kronecker.KroneckerProduct{Number,Array{Int8,2},…
```
One option is to instead pick a different wildcard, say `(:x_ο_x_ο, :y_ο_y_ο)` or `(:x_0_x_0, :y_0_y_0)`. Another would be just not to name such cases.

Reply:
> Done!
> 
> Not sure I follow, `eltype(kronecker(rand(Int32,2,2), 2)) == Int32` seems fine, but `eltype(kronecker(nda, 2)) == Number` seems like a bug?

Strange, I don't immediately see the problem here. Could merge this and add as an issue, since this will mainly impact performance?

There is no type promotion for `KroneckerPowers`, so this is truly strange...

Reply:
> The only weirdness now is that `:_` is also used as a wildcard for ordinary arrays, thus:
> 
> ```
> julia> kronecker(nda, rand(2,2), nda, rand(2,2))
> 16×16 NamedDimsArray{(:x___x__, :y___y__),Number,2,Kronecker.KroneckerProduct{Number,Array{Int8,2},…
> ```
> 
> One option is to instead pick a different wildcard, say `(:x_ο_x_ο, :y_ο_y_ο)` or `(:x_0_x_0, :y_0_y_0)`. Another would be just not to name such cases.

I would guess it mainly applies where all arrays are named? Otherwise, it makes less sense?

How about combinations of named and regular arrays always transform the regular arrays in NA with numbered rows and columns?

Reply:
The type bug is pretty odd. Maybe hold off until we sort it out. 

By "numbered rows and columns”, do you mean returning `(:x_1_x_1, :y_2_y_2)` here, always 2 for columns? (The indexing within rows/columns is unaffected by NamedDims, it just attaches `ndims` symbols, to dimensions.)

Reply:
OK I made #53 about the eltype business. I think it’s not caused by this PR.